### PR TITLE
Proj5

### PR DESCRIPTION
**Pull request by pkirlin**
Right now this PR has just some functions I've been writing that I hope will go into the main cs1 library. These functions are designed around making notebooks easier to use for the students (and for us).

There are three ok-related functions for authenticating, grading test cases, and submitting.  They all share a module-level global `_ok` variable that is used to keep track of the `Notebook` object.  This takes the place of the notebook-level `ok` object that students frequently get "undefined variable" messages about.  Because the `ok_runtests()` and `ok_submit()` functions check to make sure we are logged in first, we also don't have to use the notebook execution dependency method either.

There is also a new `reload_functions()` function.   Because python typically ignores subsequent import commands after the first one, this    function forces the specified .py file to be re-read and re-imported into the notebook namespace.  However,    any function *calls* are ignored.

This function came about for two reasons.  First, we typically have students    call `main()` at the end of their programs, and when one does `from _____ import *`    that will end up calling `main()` when all you want to do is import the functions
    into a notebook so they can be tested.  This then messes up test-case-running code, because all you want to do is mport the functions, not run `main()`, but that's not how Python works.

Furthermore, the IPython %autoreload magic    is buggy at times, and doesn't always seem to reload the functions, and when it     does, it still ends up calling `main()`.  Hence this function was designed to solve     both of those problems: it effectively executes the given Python file in the     __main__ namespace, but ignores any function calls.  Therefore, function definitions,     other imports, and global variables/constants will be imported correctly into the    main notebook space.

In a notebook cell, one can put:
```
from cs1.notebooks import *
reload_functions("student_project_file.py")
```

and then all the functions in the .py file should be available in that notebook cell, and any calls to main() or other functions in the .py file will be ignored.

PS: This functions is also an end-run around us never teaching the `if __name__ == "__main__` trick, and now we're paying the price.  :-)

### CODE REVIEW

File: projects/project05/cs1/notebooks.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    if __name__ == "__main__" trick, and now we're paying the price.  :-)
+    """
+    
+    if verbose: print("Reloading functions from", filename)
+
+    with open(filename) as f:
+        p = ast.parse(f.read(), filename=filename)
+
+        for node in p.body[:]:
+            #if not isinstance(node, ast.FunctionDef):
```
Comment (matthewlang) [inline projects/project05/cs1/notebooks.py:87]: rm

File: projects/project05/cs1/notebooks.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+            if isinstance(node, ast.Expr) and isinstance(node.value, ast.Call):   
+                p.body.remove(node)
+                if verbose: print("rejecting", node)
+            else:
+                if verbose: print("accepting", node)
+
+    # Compile the code and exec it in the __main__ namespace, which contains
+    # all the variables for the notebook cells.
+    obj = compile(p, filename=filename, mode="exec")
+    exec(obj, sys.modules["__main__"].__dict__)
```
Comment (matthewlang) [inline projects/project05/cs1/notebooks.py:100]: This is cool!

File: projects/project05/cs1/notebooks.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    _ok.auth(inline=True)
+    
+def ok_runtests(okfile, question):
+    """Run test cases and grade them using OK.
+    
+    okfile: the .ok file that describes the OK assignment we are using.
+    question: the test file to use.
+    """
+    
+    global _ok
```
Comment (matthewlang) [inline projects/project05/cs1/notebooks.py:32]: nit, take it or leave it:

```
def _maybe_login(okfile):
  global _ok
  if not _ok:
    ok_login(okfile)
```

and then call in each function.

Maybe also:
```
def ok_login(okfile):
   global _ok
   _ok = None
   _maybe_login(okfile)
```

**Review by matthewlang — COMMENTED** [review]
This is awesome! Much easier/cleaner for students. Do you want to send this to [the libs](https://github.com/Rhodes-CS-Department/comp141-libraries)?  I don't see any point in not using this right away

**Comment by pkirlin** [comment]
Now includes full project!

File: projects/project05/P5.ipynb [hunk 4]
```diff
@@ -0,0 +1,510 @@
...
+    "and they have lived at the zoo since 2003.  Ya Ya is originally from Beijing, China, while Le Le comes from Changquing, China.  The name \"Ya Ya\" means \"beautiful little girl,\" and \"Le Le,\" pronounced roughly \"luh luh,\" means \"happy happy.\"  According to the Zoo, \"...they [are] only the ninth pair in the world outside of China.  There are thought to be fewer than 1,000 giant pandas left in the wild, making it one of the most endangered species on the planet.\"  [(Source)](https://www.memphiszoo.org/blog/posts/memphis-zoo-will-continue-to-house-giant-pandas)\n",
+    "\n",
+    "[View the Memphis Zoo pandas live!](https://www.memphiszoo.org/panda-cam)\n",
+    "\n",
+    "Ya Ya and Le Le normally spend their time sleeping and eating bamboo, but occasionally they like to throw wild \n",
+    "parties in their enclosure at the zoo.  Their favorite game to play at these parties works as follows.\n",
+    "\n",
+    "Each panda is given a supply of small stones.  They take turns placing stones into a communal pile between them.  On their turn, each \n",
```
Comment (matthewlang) [inline projects/project05/P5.ipynb:29]: I think this is a little unclear -- I think a student would ask: do I need to implement a supply for each player?

Here's a stab: The pandas take turns placing small stones into a communal pile between them (assume they each have an "infinite" supply of stones to place). The game ends when the communal pile has a certain chosen maximum number of stones in it. On their turn, each...

Honestly though, this is just a nit, reading through the rest of the lab makes it clear what they need to do. Up to you 😄

**Review by matthewlang — APPROVED** [review]
This is great!

File: projects/project05/P5.ipynb (hunk 4 above)
Comment (pkirlin) [reply projects/project05/P5.ipynb:29]: Thanks, I'll edit this.

**Pull request merged**